# Raw_data_and_Bioinfo
### GC_values (data available [here](https://mega.nz/#!R6plESab!osTYeN_JhWlLOEfr5qRN4ah5d4bG9ocktPAs1wCt0i4)) 
Data generated by Nathalie Zeballos in 2018 and corrected by Sylvain Glémin & Maya Schrödl in 2019
1. SPECIES_RNA_complete:
'id': identification of RNA sequence
	'keygene': gene identification
	'nChr': name of the chromosome or scaffold on which is the sequence.
	'nbexon': number of exons per gene.
	'nbi': number of introns per gene.
2. SPECIES_exon_complete
Text file containing all information concerning CDS and UTR sequences.
	'id': identification of RNA sequence
	'keygene': gene identification
	'strand': defined as + (forward) or - (reverse) 
	'phase': One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on  
	'num': CDS item or UTR item
	'nbexon': number of exons per gene.
	'nbi': number of introns per gene.
3. SPECIES_intron_complet
	**!These files have not been corrected yet and it is possible that there are still phase problems!**
	'id' of the intron has been invented by concatening the keygene + «_intron» + 'numi' value
	'keygene' of the intron matches to 'ID' of CDS from the GFF3 file for RNA sequences. 
	'numi': intron item

### GC_values_NoMultiUTR (data available [here](https://mega.nz/#!tvBzhAbB!9haFVsoxnDVY5lPJS_C1BQ8jxnC-Tg4AtMt_-McY4e0)) 
As GC_values, but the genes containing an intron in their 3’-UTR have been deleted (as in Ressayre et al (2015, Genome Biol. Evol.), Fig. 1)
Physical_and_genetic_Maps
This folder contains the physical and genetic maps which were available. Of the four species: Oryza sativa, Brachypodium distachyon, Zea mays and Arabidopsis thaliana.
For columnname documentation see the MareyMap manual (https://cran.r-project.org/web/packages/MareyMap/vignettes/vignette.pdf)
### Phytozome_data (data available [here]())
Sequence data downloaded from Phytozome.org by Nathalie Zeballos in 2018.
### Selected_Chromosomes
Since most of the species only had scaffolds, we had to choose which scaffolds we considered as chromosomes. Furthermore, we only focused on nuclear DNA and neglected plastid/mitoDNA.
nChr: Chromosome name
lenChr: Chromosome length [bp]
### Species_Info.txt
Contains Info on each species:
species: species name abbreviation
taxonomic_group: taxonomic group
genome_size_Mb: genome size
name: full species name
### Species_List.txt
Just a textfile with all the 52 species abbreviations.
### Bioinfo_Nathalie
Analyses done by Nathalie Zeballos in 2018
